File: October 2022/Day 30/C++ Code/Chicks in a Zoo.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution {
public:
	
	// Time Complexity :- O(n)
	long long int NoOfChicks(int n){
	    // Code here
	    long long int dp[n+1][2];
        dp[1][0]=1;
        dp[1][1]=1;
        
        for(int i=2;i<=n;i++) {
            if(i>6) {
                dp[i][1]=((dp[i-1][0]-dp[i-6][1])*2);
                
                dp[i][0]=dp[i][1]+dp[i-1][0]-dp[i-6][1];
            } else {
                dp[i][1]=(dp[i-1][0]*2);
                
                dp[i][0]=dp[i][1]+dp[i-1][0];
            }
        }
        
        return dp[n][0];
	}
};

//{ Driver Code Starts.
int main(){
	int tc;
	cin >> tc;
	while(tc--){
		int N;
		cin >> N;
		Solution obj;
		long long int ans = obj.NoOfChicks(N);
		cout << ans <<"\n";
	}
	return 0;
}
// } Driver Code Ends

File: October 2022/Day 30/JAVA Code/Chicks in a Zoo.java
//{ Driver Code Starts
//Initial Template for Java

import java.util.*;
import java.lang.*;
import java.io.*;
class GFG
{
    public static void main(String[] args) throws IOException
    {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine().trim());
        while(T-->0)
        {
            int N = Integer.parseInt(br.readLine().trim());
            Solution obj = new Solution();
            long ans = obj.NoOfChicks(N);
            System.out.println(ans);
        }
    }
}

// } Driver Code Ends


//User function Template for Java

class Solution
{
    // Time Complexity :- O(n)
    public long NoOfChicks(int N)
    {
        // Code here
        long[][] dp = new long[N+1][2];
        dp[1][0]=1;
        dp[1][1]=1;
        
        for(int i=2;i<=N;i++) {
            if(i>6) {
                dp[i][1]=((dp[i-1][0]-dp[i-6][1])*2);
                
                dp[i][0]=dp[i][1]+dp[i-1][0]-dp[i-6][1];
            } else {
                dp[i][1]=(dp[i-1][0]*2);
                
                dp[i][0]=dp[i][1]+dp[i-1][0];
            }
        }
        
        return dp[N][0];
    }
}
